DemoInfo/DemoParser.cs

```diff
@@ -506,9 +527,142 @@ public void ParseHeader()
 
 			Header = header;
 
-
 			if (HeaderParsed != null)
 				HeaderParsed(this, new HeaderParsedEventArgs(Header));
+
+				if (header.PlaybackTime == 0)  //header is corrupted
+					FixHeader();
+		}
+
+		void FixHeader()
+		{
+			ExpressParseToEnd();
+			Header.PlaybackTicks = IngameTick;
+			Header.PlaybackFrames = CurrentTick - (TickAtStart + 1);
+			Header.PlaybackTime = TickInterval * IngameTick;
+			
+			CurrentTick = 0;
+
+			//the ol' switcheroo
+			//It's probably safest and easiest to
+			//swap out for a fresh bitstream.
+			Input.Seek(0, System.IO.SeekOrigin.Begin);
```

Comment:
This is potentially problematic. Consider:

* The input stream is not necessarily seekable (it could be e.g. a HTTP download). In this case, the `NotSupportedException` would only be thrown if the code coincidentally happens to hit one of these broken demos. **Very** hard to debug.
* The demo doesn't necessarily start at the beginning of the input stream. There may be some other data that the user read before handing the stream to demoinfo.

Both cases are very unlikely but still effectively break our API. My favorite solution would be to avoid the restart entirely (Can we? What are the consequences?) but if that's totally impossible, you need to verify in the constructor that the assumptions you make here about the stream actually hold. Because there is a real usecase for parsing from non-seekable streams (the heatmap website for instance), there also needs to be a flag to bypass this check and revert to today's behavior in this case.